### PR TITLE
trace: Add bootstrap tracing events

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
Support tracing when the cluster is still initializing

## Motivation and Context
To be able to understand why a cluster is so slow to start up

## How to test this PR?
Hard to test without faking the code, just add time.Sleep(10*time.Minute) before initServer()

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#a05809bb-0072-4e1b-8aab-7f969c90ed70) |
<a name="a05809bb-0072-4e1b-8aab-7f969c90ed70"></a>
### 16286-c1be640/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.2:31209
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp d28bda005063:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 1 minutes and 1 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 15 seconds
(8/14) Running minio-go tests ... done in 50 seconds
(9/14) Running minio-java tests ... FAILED in 30 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 423,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"1732D3599A895349\",\"HostId\":\"2ad59876-b7a2-4eb8-bd58-6487bf6d2295\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 51 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 46 seconds
(11/14) Running s3cmd tests ... done in 16 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 6 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/server-main.go [hunk 1]
```diff
@@ -366,6 +377,8 @@ func initServer(ctx context.Context, newObject ObjectLayer) error {
 	// lock request. This will spread locks requests over 1 second.
 	time.Sleep(time.Duration(r.Float64() * float64(time.Second)))
 
+	time.Sleep(10 * time.Second)
```
Comment (klauspost) [inline cmd/server-main.go]: Remove.

File: cmd/server-main.go [hunk 2]
```diff
@@ -374,6 +387,8 @@ func initServer(ctx context.Context, newObject ObjectLayer) error {
 		default:
 		}
 
+		globalTrace.Publish(bootstrapTrace("trying to acquire the lock"))
```
Comment (klauspost) [inline cmd/server-main.go]:
```suggestion
		globalTrace.Publish(bootstrapTrace("trying to acquire transaction.lock"))
```

File: cmd/server-main.go [hunk 3]
```diff
@@ -374,6 +387,8 @@ func initServer(ctx context.Context, newObject ObjectLayer) error {
...
+
```
Comment (klauspost) [inline cmd/server-main.go:393]: Below add something like:

```Go
		if err != nil {
			logger.Info("Waiting for all MinIO sub-systems to be initialized.. trying to acquire lock")
			
			globalTrace.Publish(bootstrapTrace(fmt.Sprintf("lock not available. error: %v. sleeping before retry", err))
```

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by klauspost — APPROVED** [review]
LGTM

File: cmd/server-main.go [hunk 4]
```diff
@@ -345,6 +346,16 @@ func configRetriableErrors(err error) bool {
 		errors.Is(err, os.ErrDeadlineExceeded)
 }
 
+func bootstrapTrace(msg string) madmin.TraceInfo {
+	return madmin.TraceInfo{
+		TraceType: madmin.TraceBootstrap,
+		Time:      time.Now().UTC(),
+		NodeName:  globalLocalNodeName,
+		FuncName:  "BOOTSTRAP",
```
Comment (harshavardhana) [inline cmd/server-main.go:358]: You can capture the line number and source @vadmeste instead of just a string.

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Just one thing rest LGTM

File: cmd/server-main.go (hunk 4 above)
Comment (vadmeste) [reply cmd/server-main.go:358]: > You can capture the line number and source @vadmeste instead of just a string.

just one line or the whole stack trace ?

Comment (klauspost) [reply cmd/server-main.go:358]: Just the func name+file:line no of the caller.

Comment (harshavardhana) [reply cmd/server-main.go:358]: yeah `"BOOTSTRAP: "+getSource(1)`  @vadmeste perhaps works?

**Review by krisis — APPROVED** [review]
LGTM

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#ae495d9e-94ca-4992-9451-fe7d079e57f2) |
<a name="ae495d9e-94ca-4992-9451-fe7d079e57f2"></a>
### 16286-1542e27/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.8:31840
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp b30a50a59181:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 1 minutes and 0 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 15 seconds
(8/14) Running minio-go tests ... done in 45 seconds
(9/14) Running minio-java tests ... FAILED in 26 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 427,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"1732E66ABC0353CE\",\"HostId\":\"539efbcb-399a-465d-9a4f-4d957f01d0cb\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 48 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 41 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :x: [more...](#d6bd22b9-d096-416c-aa20-bb6ab785b602) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#8024fb13-4f96-4c11-87c7-0cbd6985a3f0) |
<a name="8024fb13-4f96-4c11-87c7-0cbd6985a3f0"></a>
### 16286-fdb3445/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.4:30475
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp bfc72d8531af:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 7 seconds
(5/14) Running awscli tests ... done in 1 minutes and 0 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 15 seconds
(8/14) Running minio-go tests ... done in 52 seconds
(9/14) Running minio-java tests ... FAILED in 27 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 448,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"1732E747B0A103B7\",\"HostId\":\"b85a612d-e332-4d91-982d-9a2fee491d37\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 50 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 38 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
<a name="d6bd22b9-d096-416c-aa20-bb6ab785b602"></a>
### 16286-fdb3445/mint-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.3:32379
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 996487d76db8:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 42 seconds
(4/14) Running aws-sdk-ruby tests ... done in 8 seconds
(5/14) Running awscli tests ... done in 59 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 16 seconds
(8/14) Running minio-go tests ... done in 51 seconds
(9/14) Running minio-java tests ... FAILED in 28 seconds
{
  "name": "minio-java",
  "function": "addUser()",
  "duration": 363,
  "status": "FAIL",
  "error": "java.lang.RuntimeException: Request failed with response: {\"Code\":\"InternalError\",\"Message\":\"We encountered an internal error, please try again. (Server not initialized, please try again)\",\"Resource\":\"/minio/admin/v3/add-user\",\"RequestId\":\"1732E747B713A9A6\",\"HostId\":\"b683eae1-d317-4c0f-a2c0-e0b79f09ba39\"}\n >>> [io.minio.admin.MinioAdminClient.execute(MinioAdminClient.java:206), io.minio.admin.MinioAdminClient.addUser(MinioAdminClient.java:237), TestMinioAdminClient.addUser(TestMinioAdminClient.java:47), TestMinioAdminClient.runAdminTests(TestMinioAdminClient.java:160), FunctionalTest.runTests(FunctionalTest.java:3782), FunctionalTest.runEndpointTests(FunctionalTest.java:3881), FunctionalTest.main(FunctionalTest.java:4002)]"
}
(9/14) Running minio-js tests ... done in 50 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 39 seconds
(11/14) Running s3cmd tests ... done in 15 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/server-main.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/server-main.go:358]: Oh wait you already have this in message, its not needed.

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**